File: project/jimmer-sql/src/test/java/org/babyfish/jimmer/sql/dto/RecursiveTreeTest.java
package org.babyfish.jimmer.sql.dto;

import org.babyfish.jimmer.sql.common.AbstractQueryTest;
import org.babyfish.jimmer.sql.model.TreeNode;
import org.babyfish.jimmer.sql.model.TreeNodeDraft;
import org.babyfish.jimmer.sql.model.TreeNodeTable;
import org.babyfish.jimmer.sql.model.dto.RecursiveTree;
import org.junit.jupiter.api.Test;

import java.util.Arrays;
import java.util.Collections;

public class RecursiveTreeTest extends AbstractQueryTest {

    @Test
    public void testDtoToEntity() {
        RecursiveTree tree = root(
                "food",
                tree(
                        "drinks",
                        tree("cocacola"),
                        tree("fenta")
                )
        );
        assertContentEquals(
                "{" +
                        "--->\"name\":\"food\"," +
                        "--->\"childNodes\":[" +
                        "--->--->{" +
                        "--->--->--->\"name\":\"drinks\"," +
                        "--->--->--->\"childNodes\":[" +
                        "--->--->--->--->{" +
                        "--->--->--->--->--->\"name\":\"cocacola\"," +
                        "--->--->--->--->--->\"childNodes\":[]" +
                        "--->--->--->--->},{" +
                        "--->--->--->--->--->\"name\":\"fenta\"," +
                        "--->--->--->--->--->\"childNodes\":[]" +
                        "--->--->--->--->}" +
                        "--->--->--->]" +
                        "--->--->}" +
                        "--->]" +
                        "}",
                tree.toEntity().toString()
        );
    }

    @Test
    public void testEntityToDto() {
        TreeNode treeNode = TreeNodeDraft.$.produce(food -> {
           food.setName("food").addIntoChildNodes(drinks -> {
               drinks.setName("drinks");
               drinks.addIntoChildNodes(cocacola -> {
                   cocacola.setName("Cococla");
               });
               drinks.addIntoChildNodes(fenta -> {
                   fenta.setName("Fenta");
               });
           });
        });
        assertContentEquals(
                "RecursiveTree(" +
                        "--->name=food, " +
                        "--->childNodes=[" +
                        "--->--->RecursiveTree(" +
                        "--->--->--->name=drinks, " +
                        "--->--->--->childNodes=[" +
                        "--->--->--->--->RecursiveTree(" +
                        "--->--->--->--->--->name=Cococla, " +
                        "--->--->--->--->--->childNodes=null" +
                        "--->--->--->--->), " +
                        "--->--->--->--->RecursiveTree(" +
                        "--->--->--->--->--->name=Fenta, " +
                        "--->--->--->--->--->childNodes=null" +
                        "--->--->--->--->)" +
                        "--->--->--->]" +
                        "--->--->)" +
                        "--->]" +
                        ")",
                new RecursiveTree(treeNode)
        );
    }

    @Test
    public void testEntityToDto2() {
        TreeNode treeNode = TreeNodeDraft.$.produce(food -> {
            food.setName("food").addIntoChildNodes(drinks -> {
                drinks.setName("drinks");
                drinks.addIntoChildNodes(cocacola -> {
                    cocacola.setName("Cococla");
                    cocacola.setChildNodes(Collections.emptyList());
                });
                drinks.addIntoChildNodes(fenta -> {
                    fenta.setName("Fenta");
                    fenta.setChildNodes(Collections.emptyList());
                });
            });
        });
        assertContentEquals(
                "RecursiveTree(" +
                        "--->name=food, " +
                        "--->childNodes=[" +
                        "--->--->RecursiveTree(" +
                        "--->--->--->name=drinks, " +
                        "--->--->--->childNodes=[" +
                        "--->--->--->--->RecursiveTree(" +
                        "--->--->--->--->--->name=Cococla, " +
                        "--->--->--->--->--->childNodes=[]" +
                        "--->--->--->--->), " +
                        "--->--->--->--->RecursiveTree(" +
                        "--->--->--->--->--->name=Fenta, " +
                        "--->--->--->--->--->childNodes=[]" +
                        "--->--->--->--->)" +
                        "--->--->--->]" +
                        "--->--->)" +
                        "--->]" +
                        ")",
                new RecursiveTree(treeNode)
        );
    }

    @Test
    public void findFindTree() {

        TreeNodeTable table = TreeNodeTable.$;

        executeAndExpect(
                getSqlClient()
                        .createQuery(table)
                        .where(table.parentId().isNull())
                        .select(table.fetch(RecursiveTree.class)),
                ctx -> {
                    ctx.sql(
                            "select tb_1_.NODE_ID, tb_1_.NAME " +
                                    "from TREE_NODE tb_1_ " +
                                    "where tb_1_.PARENT_ID is null"
                    );
                    ctx.statement(1).sql(
                            "select tb_1_.NODE_ID, tb_1_.NAME " +
                                    "from TREE_NODE tb_1_ " +
                                    "where tb_1_.PARENT_ID = ? " +
                                    "order by tb_1_.NODE_ID asc"
                    );
                    ctx.statement(2).sql(
                            "select tb_1_.PARENT_ID, tb_1_.NODE_ID, tb_1_.NAME " +
                                    "from TREE_NODE tb_1_ " +
                                    "where tb_1_.PARENT_ID in (?, ?) " +
                                    "order by tb_1_.NODE_ID asc"
                    );
                    ctx.statement(3).sql(
                            "select tb_1_.PARENT_ID, tb_1_.NODE_ID, tb_1_.NAME " +
                                    "from TREE_NODE tb_1_ " +
                                    "where tb_1_.PARENT_ID in (?, ?, ?, ?) " +
                                    "order by tb_1_.NODE_ID asc"
                    );
                    ctx.statement(4).sql(
                            "select tb_1_.PARENT_ID, tb_1_.NODE_ID, tb_1_.NAME " +
                                    "from TREE_NODE tb_1_ " +
                                    "where tb_1_.PARENT_ID in (?, ?, ?, ?, ?, ?, ?, ?) " +
                                    "order by tb_1_.NODE_ID asc"
                    );
                    ctx.statement(5).sql(
                            "select tb_1_.PARENT_ID, tb_1_.NODE_ID, tb_1_.NAME " +
                                    "from TREE_NODE tb_1_ " +
                                    "where tb_1_.PARENT_ID in (?, ?, ?, ?, ?, ?, ?, ?, ?) " +
                                    "order by tb_1_.NODE_ID asc"
                    );
                    ctx.rows(
                            "[" +
                                    "--->{" +
                                    "--->--->\"name\":\"Home\"," +
                                    "--->--->\"childNodes\":[" +
                                    "--->--->--->{" +
                                    "--->--->--->--->\"name\":\"Food\"," +
                                    "--->--->--->--->\"recursiveChildNodes\":[" +
                                    "--->--->--->--->--->{" +
                                    "--->--->--->--->--->--->\"name\":\"Drinks\"," +
                                    "--->--->--->--->--->--->\"recursiveChildNodes\":[" +
                                    "--->--->--->--->--->--->--->{" +
                                    "--->--->--->--->--->--->--->--->\"name\":\"Coca Cola\"," +
                                    "--->--->--->--->--->--->--->--->\"recursiveChildNodes\":[" +
                                    "--->--->--->--->--->--->--->--->]" +
                                    "--->--->--->--->--->--->--->}," +
                                    "--->--->--->--->--->--->--->{" +
                                    "--->--->--->--->--->--->--->--->\"name\":\"Fanta\"," +
                                    "--->--->--->--->--->--->--->--->\"recursiveChildNodes\":[" +
                                    "--->--->--->--->--->--->--->--->]" +
                                    "--->--->--->--->--->--->--->}" +
                                    "--->--->--->--->--->--->]" +
                                    "--->--->--->--->--->}," +
                                    "--->--->--->--->--->{" +
                                    "--->--->--->--->--->--->\"name\":\"Bread\"," +
                                    "--->--->--->--->--->--->\"recursiveChildNodes\":[" +
                                    "--->--->--->--->--->--->--->{" +
                                    "--->--->--->--->--->--->--->--->\"name\":\"Baguette\"," +
                                    "--->--->--->--->--->--->--->--->\"recursiveChildNodes\":[" +
                                    "--->--->--->--->--->--->--->--->]" +
                                    "--->--->--->--->--->--->--->}," +
                                    "--->--->--->--->--->--->--->{" +
                                    "--->--->--->--->--->--->--->--->\"name\":\"Ciabatta\"," +
                                    "--->--->--->--->--->--->--->--->\"recursiveChildNodes\":[" +
                                    "--->--->--->--->--->--->--->--->]" +
                                    "--->--->--->--->--->--->--->}" +
                                    "--->--->--->--->--->--->]" +
                                    "--->--->--->--->--->}" +
                                    "--->--->--->--->]" +
                                    "--->--->--->}," +
                                    "--->--->--->{" +
                                    "--->--->--->--->\"name\":\"Clothing\"," +
                                    "--->--->--->--->\"recursiveChildNodes\":[" +
                                    "--->--->--->--->--->{" +
                                    "--->--->--->--->--->--->\"name\":\"Woman\"," +
                                    "--->--->--->--->--->--->\"recursiveChildNodes\":[" +
                                    "--->--->--->--->--->--->--->{" +
                                    "--->--->--->--->--->--->--->--->\"name\":\"Casual wear\"," +
                                    "--->--->--->--->--->--->--->--->\"recursiveChildNodes\":[" +
                                    "--->--->--->--->--->--->--->--->--->{" +
                                    "--->--->--->--->--->--->--->--->--->--->\"name\":\"Dress\"," +
                                    "--->--->--->--->--->--->--->--->--->--->\"recursiveChildNodes\":[" +
                                    "--->--->--->--->--->--->--->--->--->--->]" +
                                    "--->--->--->--->--->--->--->--->--->}," +
                                    "--->--->--->--->--->--->--->--->--->{" +
                                    "--->--->--->--->--->--->--->--->--->--->\"name\":\"Miniskirt\"," +
                                    "--->--->--->--->--->--->--->--->--->--->\"recursiveChildNodes\":[" +
                                    "--->--->--->--->--->--->--->--->--->--->]" +
                                    "--->--->--->--->--->--->--->--->--->}," +
                                    "--->--->--->--->--->--->--->--->--->{" +
                                    "--->--->--->--->--->--->--->--->--->--->\"name\":\"Jeans\"," +
                                    "--->--->--->--->--->--->--->--->--->--->\"recursiveChildNodes\":[" +
                                    "--->--->--->--->--->--->--->--->--->--->]" +
                                    "--->--->--->--->--->--->--->--->--->}" +
                                    "--->--->--->--->--->--->--->--->]" +
                                    "--->--->--->--->--->--->--->}," +
                                    "--->--->--->--->--->--->--->{" +
                                    "--->--->--->--->--->--->--->--->\"name\":\"Formal wear\"," +
                                    "--->--->--->--->--->--->--->--->\"recursiveChildNodes\":[" +
                                    "--->--->--->--->--->--->--->--->--->{" +
                                    "--->--->--->--->--->--->--->--->--->--->\"name\":\"Suit\"," +
                                    "--->--->--->--->--->--->--->--->--->--->\"recursiveChildNodes\":[" +
                                    "--->--->--->--->--->--->--->--->--->--->]" +
                                    "--->--->--->--->--->--->--->--->--->}," +
                                    "--->--->--->--->--->--->--->--->--->{" +
                                    "--->--->--->--->--->--->--->--->--->--->\"name\":\"Shirt\"," +
                                    "--->--->--->--->--->--->--->--->--->--->\"recursiveChildNodes\":[" +
                                    "--->--->--->--->--->--->--->--->--->--->]" +
                                    "--->--->--->--->--->--->--->--->--->}" +
                                    "--->--->--->--->--->--->--->--->]" +
                                    "--->--->--->--->--->--->--->}" +
                                    "--->--->--->--->--->--->]" +
                                    "--->--->--->--->--->}," +
                                    "--->--->--->--->--->{" +
                                    "--->--->--->--->--->--->\"name\":\"Man\"," +
                                    "--->--->--->--->--->--->\"recursiveChildNodes\":[" +
                                    "--->--->--->--->--->--->--->{" +
                                    "--->--->--->--->--->--->--->--->\"name\":\"Casual wear\"," +
                                    "--->--->--->--->--->--->--->--->\"recursiveChildNodes\":[" +
                                    "--->--->--->--->--->--->--->--->--->{" +
                                    "--->--->--->--->--->--->--->--->--->--->\"name\":\"Jacket\"," +
                                    "--->--->--->--->--->--->--->--->--->--->\"recursiveChildNodes\":[" +
                                    "--->--->--->--->--->--->--->--->--->--->]" +
                                    "--->--->--->--->--->--->--->--->--->}," +
                                    "--->--->--->--->--->--->--->--->--->{" +
                                    "--->--->--->--->--->--->--->--->--->--->\"name\":\"Jeans\"," +
                                    "--->--->--->--->--->--->--->--->--->--->\"recursiveChildNodes\":[" +
                                    "--->--->--->--->--->--->--->--->--->--->]" +
                                    "--->--->--->--->--->--->--->--->--->}" +
                                    "--->--->--->--->--->--->--->--->]" +
                                    "--->--->--->--->--->--->--->}," +
                                    "--->--->--->--->--->--->--->{" +
                                    "--->--->--->--->--->--->--->--->\"name\":\"Formal wear\"," +
                                    "--->--->--->--->--->--->--->--->\"recursiveChildNodes\":[" +
                                    "--->--->--->--->--->--->--->--->--->{" +
                                    "--->--->--->--->--->--->--->--->--->--->\"name\":\"Suit\"," +
                                    "--->--->--->--->--->--->--->--->--->--->\"recursiveChildNodes\":[" +
                                    "--->--->--->--->--->--->--->--->--->--->]" +
                                    "--->--->--->--->--->--->--->--->--->}," +
                                    "--->--->--->--->--->--->--->--->--->{" +
                                    "--->--->--->--->--->--->--->--->--->--->\"name\":\"Shirt\"," +
                                    "--->--->--->--->--->--->--->--->--->--->\"recursiveChildNodes\":[" +
                                    "--->--->--->--->--->--->--->--->--->--->]" +
                                    "--->--->--->--->--->--->--->--->--->}" +
                                    "--->--->--->--->--->--->--->--->]" +
                                    "--->--->--->--->--->--->--->}" +
                                    "--->--->--->--->--->--->]" +
                                    "--->--->--->--->--->}" +
                                    "--->--->--->--->]" +
                                    "--->--->--->}" +
                                    "--->--->]" +
                                    "--->}" +
                                    "]--->"
                    );
                }
        );
    }

    private static RecursiveTree root(String name, RecursiveTree ...childNodes) {
        RecursiveTree tree = new RecursiveTree();
        tree.setName(name);
        if (childNodes.length > 0) {
            tree.setRecursiveChildNodes(Arrays.asList(childNodes));
        }
        return tree;
    }

    private static RecursiveTree tree(String name, RecursiveTree ...childNodes) {
        RecursiveTree tree = new RecursiveTree();
        tree.setName(name);
        if (childNodes.length > 0) {
            tree.setRecursiveChildNodes(Arrays.asList(childNodes));
        }
        return tree;
    }
}

File: project/jimmer-sql-kotlin/src/test/kotlin/org/babyfish/jimmer/sql/kt/dto/RecursiveTreeTest.kt
package org.babyfish.jimmer.sql.kt.dto

import org.babyfish.jimmer.kt.new
import org.babyfish.jimmer.sql.kt.ast.expression.isNull
import org.babyfish.jimmer.sql.kt.common.AbstractQueryTest
import org.babyfish.jimmer.sql.kt.common.assertContent
import org.babyfish.jimmer.sql.kt.model.TreeNode
import org.babyfish.jimmer.sql.kt.model.addBy
import org.babyfish.jimmer.sql.kt.model.by
import org.babyfish.jimmer.sql.kt.model.dto.RecursiveTree
import org.babyfish.jimmer.sql.kt.model.parentId
import org.junit.Test

class RecursiveTreeTest : AbstractQueryTest() {

    @Test
    fun testDtoToEntity() {
        val tree = RecursiveTree(
            name = "food",
            childNodes = listOf(
                RecursiveTree(
                    name = "drinks",
                    childNodes = listOf(
                        RecursiveTree(
                            name = "cocacola",
                            childNodes = emptyList()
                        ),
                        RecursiveTree(
                            name = "fenta",
                            childNodes = emptyList()
                        )
                    )
                )
            )
        )

        assertContent(
            """{
                |--->"name":"food",
                |--->"childNodes":[
                |--->--->{
                |--->--->--->"name":"drinks",
                |--->--->--->"childNodes":[
                |--->--->--->--->{"name":"cocacola","childNodes":[]}
                |--->--->--->--->,{"name":"fenta","childNodes":[]}
                |--->--->--->]
                |--->--->}
                |--->]
                |}""".trimMargin(),
            tree.toEntity()
        )
    }

    @Test
    fun testEntityToDto() {
        val treeNode = new(TreeNode::class).by {
            name = "food"
            childNodes().addBy {
                name = "drinks"
                childNodes().addBy {
                    name = "cocacola"
                }
                childNodes().addBy {
                    name = "fenta"
                }
            }
        }
        assertContent(
            "RecursiveTree(" +
                "--->name=food, " +
                "--->childNodes=[" +
                "--->--->RecursiveTree(" +
                "--->--->--->name=drinks, " +
                "--->--->--->childNodes=[" +
                "--->--->--->--->RecursiveTree(name=cocacola, childNodes=null), " +
                "--->--->--->--->RecursiveTree(name=fenta, childNodes=null)" +
                "--->--->--->]" +
                "--->--->)" +
                "--->]" +
                ")",
            RecursiveTree(treeNode)
        )
    }

    @Test
    fun testEntityToDto2() {
        val treeNode = new(TreeNode::class).by {
            name = "food"
            childNodes().addBy {
                name = "drinks"
                childNodes().addBy {
                    name = "cocacola"
                    childNodes = emptyList()
                }
                childNodes().addBy {
                    name = "fenta"
                    childNodes = emptyList()
                }
            }
        }
        assertContent(
            "RecursiveTree(" +
                "--->name=food, " +
                "--->childNodes=[" +
                "--->--->RecursiveTree(" +
                "--->--->--->name=drinks, " +
                "--->--->--->childNodes=[" +
                "--->--->--->--->RecursiveTree(name=cocacola, childNodes=[]), " +
                "--->--->--->--->RecursiveTree(name=fenta, childNodes=[])" +
                "--->--->--->]" +
                "--->--->)" +
                "--->]" +
                ")",
            RecursiveTree(treeNode)
        )
    }

    @Test
    fun testFindTree() {
        executeAndExpect(
            sqlClient.createQuery(TreeNode::class) {
                where(table.parentId.isNull())
                select(table.fetch(RecursiveTree::class))
            }
        ) {
            sql(
                "select tb_1_.NODE_ID, tb_1_.NAME " +
                    "from TREE_NODE tb_1_ " +
                    "where tb_1_.PARENT_ID is null"
            )
            statement(1).sql(
                "select tb_1_.NODE_ID, tb_1_.NAME " +
                    "from TREE_NODE tb_1_ " +
                    "where tb_1_.PARENT_ID = ? " +
                    "order by tb_1_.NODE_ID asc"
            )
            statement(2).sql(
                "select tb_1_.PARENT_ID, tb_1_.NODE_ID, tb_1_.NAME " +
                    "from TREE_NODE tb_1_ " +
                    "where tb_1_.PARENT_ID in (?, ?) " +
                    "order by tb_1_.NODE_ID asc"
            )
            statement(3).sql(
                "select tb_1_.PARENT_ID, tb_1_.NODE_ID, tb_1_.NAME " +
                    "from TREE_NODE tb_1_ " +
                    "where tb_1_.PARENT_ID in (?, ?, ?, ?) " +
                    "order by tb_1_.NODE_ID asc"
            )
            statement(4).sql(
                "select tb_1_.PARENT_ID, tb_1_.NODE_ID, tb_1_.NAME " +
                    "from TREE_NODE tb_1_ " +
                    "where tb_1_.PARENT_ID in (?, ?, ?, ?, ?, ?, ?, ?) " +
                    "order by tb_1_.NODE_ID asc"
            )
            statement(5).sql(
                "select tb_1_.PARENT_ID, tb_1_.NODE_ID, tb_1_.NAME " +
                    "from TREE_NODE tb_1_ " +
                    "where tb_1_.PARENT_ID in (?, ?, ?, ?, ?, ?, ?, ?, ?) " +
                    "order by tb_1_.NODE_ID asc"
            )
            rows(
                "[" +
                    "--->{" +
                    "--->--->\"name\":\"Home\"," +
                    "--->--->\"childNodes\":[" +
                    "--->--->--->{" +
                    "--->--->--->--->\"name\":\"Food\"," +
                    "--->--->--->--->\"childNodes\":[" +
                    "--->--->--->--->--->{" +
                    "--->--->--->--->--->--->\"name\":\"Drinks\"," +
                    "--->--->--->--->--->--->\"childNodes\":[" +
                    "--->--->--->--->--->--->--->{" +
                    "--->--->--->--->--->--->--->--->\"name\":\"Coca Cola\"," +
                    "--->--->--->--->--->--->--->--->\"childNodes\":[" +
                    "--->--->--->--->--->--->--->--->]" +
                    "--->--->--->--->--->--->--->}," +
                    "--->--->--->--->--->--->--->{" +
                    "--->--->--->--->--->--->--->--->\"name\":\"Fanta\"," +
                    "--->--->--->--->--->--->--->--->\"childNodes\":[" +
                    "--->--->--->--->--->--->--->--->]" +
                    "--->--->--->--->--->--->--->}" +
                    "--->--->--->--->--->--->]" +
                    "--->--->--->--->--->}," +
                    "--->--->--->--->--->{" +
                    "--->--->--->--->--->--->\"name\":\"Bread\"," +
                    "--->--->--->--->--->--->\"childNodes\":[" +
                    "--->--->--->--->--->--->--->{" +
                    "--->--->--->--->--->--->--->--->\"name\":\"Baguette\"," +
                    "--->--->--->--->--->--->--->--->\"childNodes\":[" +
                    "--->--->--->--->--->--->--->--->]" +
                    "--->--->--->--->--->--->--->}," +
                    "--->--->--->--->--->--->--->{" +
                    "--->--->--->--->--->--->--->--->\"name\":\"Ciabatta\"," +
                    "--->--->--->--->--->--->--->--->\"childNodes\":[" +
                    "--->--->--->--->--->--->--->--->]" +
                    "--->--->--->--->--->--->--->}" +
                    "--->--->--->--->--->--->]" +
                    "--->--->--->--->--->}" +
                    "--->--->--->--->]" +
                    "--->--->--->}," +
                    "--->--->--->{" +
                    "--->--->--->--->\"name\":\"Clothing\"," +
                    "--->--->--->--->\"childNodes\":[" +
                    "--->--->--->--->--->{" +
                    "--->--->--->--->--->--->\"name\":\"Woman\"," +
                    "--->--->--->--->--->--->\"childNodes\":[" +
                    "--->--->--->--->--->--->--->{" +
                    "--->--->--->--->--->--->--->--->\"name\":\"Casual wear\"," +
                    "--->--->--->--->--->--->--->--->\"childNodes\":[" +
                    "--->--->--->--->--->--->--->--->--->{" +
                    "--->--->--->--->--->--->--->--->--->--->\"name\":\"Dress\"," +
                    "--->--->--->--->--->--->--->--->--->--->\"childNodes\":[" +
                    "--->--->--->--->--->--->--->--->--->--->]" +
                    "--->--->--->--->--->--->--->--->--->}," +
                    "--->--->--->--->--->--->--->--->--->{" +
                    "--->--->--->--->--->--->--->--->--->--->\"name\":\"Miniskirt\"," +
                    "--->--->--->--->--->--->--->--->--->--->\"childNodes\":[" +
                    "--->--->--->--->--->--->--->--->--->--->]" +
                    "--->--->--->--->--->--->--->--->--->}," +
                    "--->--->--->--->--->--->--->--->--->{" +
                    "--->--->--->--->--->--->--->--->--->--->\"name\":\"Jeans\"," +
                    "--->--->--->--->--->--->--->--->--->--->\"childNodes\":[" +
                    "--->--->--->--->--->--->--->--->--->--->]" +
                    "--->--->--->--->--->--->--->--->--->}" +
                    "--->--->--->--->--->--->--->--->]" +
                    "--->--->--->--->--->--->--->}," +
                    "--->--->--->--->--->--->--->{" +
                    "--->--->--->--->--->--->--->--->\"name\":\"Formal wear\"," +
                    "--->--->--->--->--->--->--->--->\"childNodes\":[" +
                    "--->--->--->--->--->--->--->--->--->{" +
                    "--->--->--->--->--->--->--->--->--->--->\"name\":\"Suit\"," +
                    "--->--->--->--->--->--->--->--->--->--->\"childNodes\":[" +
                    "--->--->--->--->--->--->--->--->--->--->]" +
                    "--->--->--->--->--->--->--->--->--->}," +
                    "--->--->--->--->--->--->--->--->--->{" +
                    "--->--->--->--->--->--->--->--->--->--->\"name\":\"Shirt\"," +
                    "--->--->--->--->--->--->--->--->--->--->\"childNodes\":[" +
                    "--->--->--->--->--->--->--->--->--->--->]" +
                    "--->--->--->--->--->--->--->--->--->}" +
                    "--->--->--->--->--->--->--->--->]" +
                    "--->--->--->--->--->--->--->}" +
                    "--->--->--->--->--->--->]" +
                    "--->--->--->--->--->}," +
                    "--->--->--->--->--->{" +
                    "--->--->--->--->--->--->\"name\":\"Man\"," +
                    "--->--->--->--->--->--->\"childNodes\":[" +
                    "--->--->--->--->--->--->--->{" +
                    "--->--->--->--->--->--->--->--->\"name\":\"Casual wear\"," +
                    "--->--->--->--->--->--->--->--->\"childNodes\":[" +
                    "--->--->--->--->--->--->--->--->--->{" +
                    "--->--->--->--->--->--->--->--->--->--->\"name\":\"Jacket\"," +
                    "--->--->--->--->--->--->--->--->--->--->\"childNodes\":[" +
                    "--->--->--->--->--->--->--->--->--->--->]" +
                    "--->--->--->--->--->--->--->--->--->}," +
                    "--->--->--->--->--->--->--->--->--->{" +
                    "--->--->--->--->--->--->--->--->--->--->\"name\":\"Jeans\"," +
                    "--->--->--->--->--->--->--->--->--->--->\"childNodes\":[" +
                    "--->--->--->--->--->--->--->--->--->--->]" +
                    "--->--->--->--->--->--->--->--->--->}" +
                    "--->--->--->--->--->--->--->--->]" +
                    "--->--->--->--->--->--->--->}," +
                    "--->--->--->--->--->--->--->{" +
                    "--->--->--->--->--->--->--->--->\"name\":\"Formal wear\"," +
                    "--->--->--->--->--->--->--->--->\"childNodes\":[" +
                    "--->--->--->--->--->--->--->--->--->{" +
                    "--->--->--->--->--->--->--->--->--->--->\"name\":\"Suit\"," +
                    "--->--->--->--->--->--->--->--->--->--->\"childNodes\":[" +
                    "--->--->--->--->--->--->--->--->--->--->]" +
                    "--->--->--->--->--->--->--->--->--->}," +
                    "--->--->--->--->--->--->--->--->--->{" +
                    "--->--->--->--->--->--->--->--->--->--->\"name\":\"Shirt\"," +
                    "--->--->--->--->--->--->--->--->--->--->\"childNodes\":[" +
                    "--->--->--->--->--->--->--->--->--->--->]" +
                    "--->--->--->--->--->--->--->--->--->}" +
                    "--->--->--->--->--->--->--->--->]" +
                    "--->--->--->--->--->--->--->}" +
                    "--->--->--->--->--->--->]" +
                    "--->--->--->--->--->}" +
                    "--->--->--->--->]" +
                    "--->--->--->}" +
                    "--->--->]" +
                    "--->}" +
                    "]"
            )
        }
    }
}